src/credentials/FabricTable.cpp

```diff
@@ -1248,97 +1383,548 @@ CHIP_ERROR FabricTable::AllocatePendingOperationalKey(Optional<FabricIndex> fabr
+                                                  Crypto::P256Keypair * existingOpKey, bool isExistingOpKeyExternallyOwned,
+                                                  FabricIndex * outNewFabricIndex)
+{
+    VerifyOrReturnError(mOpCertStore != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(outNewFabricIndex != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    static_assert(kMaxValidFabricIndex <= UINT8_MAX, "Cannot create more fabrics than UINT8_MAX");
+
+    // We should already have a pending root when we get here
+    VerifyOrReturnError(mStateFlags.Has(StateFlags::kIsTrustedRootPending), CHIP_ERROR_INCORRECT_STATE);
+    // We should not have pending update when we get here
+    VerifyOrReturnError(!mStateFlags.Has(StateFlags::kIsUpdatePending), CHIP_ERROR_INCORRECT_STATE);
+
+    EnsureNextAvailableFabricIndexUpdated();
+    FabricIndex fabricIndexToUse = kUndefinedFabricIndex;
+    if (mNextAvailableFabricIndex.HasValue())
+    {
+        // Mark we have some pending data for a given fabric.
+        fabricIndexToUse = mNextAvailableFabricIndex.Value();
+    }
+    else
+    {
+        // Fabric addition, but adding fabric would fail on table full: let's not allocate a fabric
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    if (existingOpKey == nullptr)
+    {
+        // If existing operational key not provided, we need to have a keystore present.
+        // It should already have an operational key pending.
+        VerifyOrReturnError(mOperationalKeystore != nullptr, CHIP_ERROR_KEY_NOT_FOUND);
+        // Make sure we have an operational key, pending or not
+        VerifyOrReturnError(mOperationalKeystore->HasOpKeypairForFabric(fabricIndexToUse) ||
+                                mOperationalKeystore->HasPendingOpKeypair(),
+                            CHIP_ERROR_KEY_NOT_FOUND);
+    }
+
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndexToUse), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    // Check for new fabric colliding with an existing fabric
+    if (!mStateFlags.Has(StateFlags::kAreCollidingFabricsIgnored))
+    {
+        FabricIndex collidingFabricIndex = kUndefinedFabricIndex;
+        ReturnErrorOnFailure(FindExistingFabricByNocChaining(fabricIndexToUse, noc, collidingFabricIndex));
+        ReturnErrorCodeIf(collidingFabricIndex != kUndefinedFabricIndex, CHIP_ERROR_FABRIC_EXISTS);
+    }
+
+    // We don't have a collision, handle the temp insert of NOC/ICAC
+    ReturnErrorOnFailure(mOpCertStore->AddNewOpCertsForFabric(fabricIndexToUse, noc, icac));
+    VerifyOrReturnError(SetPendingDataFabricIndex(fabricIndexToUse), CHIP_ERROR_INCORRECT_STATE);
+
+    CHIP_ERROR err =
+        AddOrUpdateInner(kUndefinedFabricIndex, existingOpKey, isExistingOpKeyExternallyOwned, vendorId, outNewFabricIndex);
+    if (err != CHIP_NO_ERROR)
+    {
+        // Revert partial state added on error
+        mOpCertStore->RevertPendingOpCertsExceptRoot();
```

Comment:
Should this also undo whatever `SetPendingDataFabricIndex` did?

Reply:
I don't see how.  `SetPendingDataFabricIndex` sets members of `FabricTable`.

Reply:
Added a comment.

SetPendingDataFabricIndex just touches mFabricIndexWithPendingState. `RevertPendingOpCertsExceptRoot`.

RevertPendingOpCertsExceptRoot does:
```
    if (!mStateFlags.Has(StateFlags::kIsTrustedRootPending))
    {
        mFabricIndexWithPendingState = kUndefinedFabricIndex;
    }
```

That state revert is fine. There is nothing else to revert in that UpdateCommon case

Reply:
> RevertPendingOpCertsExceptRoot does:

That's `FabricTable::RevertPendingOpCertsExceptRoot`.  We're not calling that here.  We're calling `mOpCertStore->RevertPendingOpCertsExceptRoot()`.

Reply:
Gotcha, will do in the follow-up, as this doesn't get hit unless everything else is broken in Add/UpdateNOC